File: LeetCodePractice/src/BitManipulation/SingleNumberIII.java
package BitManipulation;

import java.util.*;

/*Given 2*n + 2 numbers, every numbers occurs twice except two, find them.

Example
Example 1:
	Input:  [1,2,2,3,4,4,5,3]
	Output:  [1,5]

Example 2:
	Input: [1,1,2,3,4,4]
	Output:  [2,3]
	
Challenge
O(n) time, O(1) extra space.*/
public class SingleNumberIII {
	public List<Integer> singleNumberIII(int[] A) {
        int xor = 0;
        for (int i = 0; i < A.length; i++) {
            xor ^= A[i];
        }
        
        int lastBit = xor - (xor & (xor - 1));
        int group0 = 0, group1 = 0;
        for (int i = 0; i < A.length; i++) {
            if ((lastBit & A[i]) == 0) {
                group0 ^= A[i];
            } else {
                group1 ^= A[i];
            }
        }
        
        ArrayList<Integer> result = new ArrayList<Integer>();
        result.add(group0);
        result.add(group1);
        return result;
    }
}

File: LeetCodePython/SingleNumberIII.py
"""
Given an integer array nums, in which exactly two elements appear only once and all the other elements appear exactly twice. Find the two elements that appear only once. You can return the answer in any order.

You must write an algorithm that runs in linear runtime complexity and uses only constant extra space.

Example 1:
Input: nums = [1,2,1,3,2,5]
Output: [3,5]
Explanation:  [5, 3] is also a valid answer.

Example 2:
Input: nums = [-1,0]
Output: [-1,0]
Example 3:

Input: nums = [0,1]
Output: [1,0]

Constraints:
2 <= nums.length <= 3 * 104
-231 <= nums[i] <= 231 - 1
Each integer in nums will appear twice, only two integers will appear once.
"""
from typing import List


class SingleNumberIII:
    def singleNumber(self, nums: List[int]) -> List[int]:
        xor, group0, group1 = 0, 0, 0
        for n in nums:
            xor ^= n
        last_bit = xor & -xor
        for n in nums:
            if (last_bit & n) == 0:
                group0 ^= n
            else:
                group1 ^= n
        return [group0, group1]
